ostd/src/sync/wait.rs

```diff
@@ -227,6 +210,40 @@ impl Waiter {
         self.waker.do_wait();
     }
 
+    /// Waits until some condition is met or the cancel condition becomes true.
+    ///
+    /// This method will return `Some(_)` if the condition returns `Some(_)`, and will return
+    /// the condition test result regardless what it is when the cancel condition becomes true.
+    #[doc(hidden)]
+    pub fn wait_until_or_cancelled<F, R, FCancel>(
+        &self,
+        mut cond: F,
+        cancel_cond: FCancel,
+        wait_queue: Option<&WaitQueue>,
+    ) -> Option<R>
+    where
+        F: FnMut() -> Option<R>,
+        FCancel: Fn() -> bool,
+    {
+        loop {
+            if let Some(wait_queue) = wait_queue {
+                wait_queue.enqueue(self.waker());
+            }
```

Comment:
This is a compact way. Yet I wonder whether `enqueue` method can be public? I think it may be a little bit weird since it may lead to some waker that is not belonged to the waitqueue, but I do need `enqueue` in aster-nix so that I can use Waiter::wait_until_or_timeout, Waiter::pause_until... to implement the corresponding WaitQueue methods directly.

Another way  is defining a method to `WaitQueue` like:
```rust
    pub fn wait_with_enqueue<F, R, Op, E, FInner>(&self, waiter: Waiter, mut cond: F, wait_operate: Op) -> 
    core::result::Result<R, E> 
    where 
        F: FnMut() -> Option<R>,
        Op: FnOnce(Waiter, ...) -> core::result::Result<R, E>,
        FInner: FnMut() -> Option<R>,
    {
        let cond = || {
            self.enqueue(waiter.waker());
            cond()
        };

        wait_operate(waiter, cond)
    }
```
But this seems cannot work because I cannot put a closure as a parameter in `FnOnce`.

Reply:
> This is a compact way. Yet I wonder whether enqueue method can be public? I think it may be a little bit weird since it may lead to some waker that is not belonged to the waitqueue, but I do need enqueue in aster-nix so that I can use Waiter::wait_until_or_timeout, Waiter::pause_until... to implement the corresponding WaitQueue methods directly.

Well, that seems to be correct. I think it is acceptable to mark it as `pub` and `#doc(hidden)`, at least for now. We can think about if there are other means in the future, I could not think of a better way at this moment.

However, I think we can remove `#[doc(hidden)]` from `Waiter::wait_until_or_canceled`. If I remember correctly, @StevenJiang1110's original concerns motivating the `#[doc(hidden)]` marker were as follows:
 - Using `Waiter` to implement `WaitQueue` is kind of an implementation detail. Now this argument becomes invalid because the method belongs to `Waiter`.
   - This argument is still valid for the `WaitQueue::enqueue` method, so we mark it as `#[doc(hidden)]`, as described above.
 - It is not clear if we have more than one user of this method. Now it's clear that we will have at least two users: one for timeout and one for signaling.